fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/api/event/player/BlockBreakEvents.java

```diff
@@ -0,0 +1,122 @@
+ */
+
+package net.fabricmc.fabric.api.event.player;
+
+import net.minecraft.block.BlockState;
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.util.ActionResult;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+
+public final class BlockBreakEvents {
+	private BlockBreakEvents() { }
+
+	/**
+	 * Callback before a block is broken.
+	 * Only called on the server, however updates are synced with the client.
+	 *
+	 * <p>Upon return:
+	 * <ul><li>SUCCESS/PASS passes on the callback to the next listener
+	 * <li>CONSUME continues the default code for breaking the block and ignores all other listeners
+	 * <li>FAIL cancels the block breaking action</ul>
+	 */
+	public static final Event<Before> BEFORE = EventFactory.createArrayBacked(Before.class,
+			(listeners) -> (world, player, pos, state, entity) -> {
+				for (Before event : listeners) {
+					ActionResult result = event.beforeBlockBreak(world, player, pos, state, entity);
+
+					if (result != ActionResult.PASS) {
+						return result;
+					}
+				}
+
+				return ActionResult.PASS;
+			}
+	);
+
+	/*
+	 * Callback after a block is broken.
+	 * Called on the Server only
+	 */
+	public static final Event<After> AFTER = EventFactory.createArrayBacked(After.class,
+			(listeners) -> (world, player, pos, state, entity) -> {
+				for (After event : listeners) {
+					event.afterBlockBreak(world, player, pos, state, entity);
+				}
+			}
+	);
+
+	/*
+	 * Callback when a block break has been canceled
+	 * Called on the Server only
+	 */
+	public static final Event<Cancel> CANCEL = EventFactory.createArrayBacked(Cancel.class,
+			(listeners) -> (world, player, pos, state, entity) -> {
+				for (Cancel event : listeners) {
+					event.onBlockBreakCancel(world, player, pos, state, entity);
+				}
+			}
+	);
+
+	@FunctionalInterface
+	public interface Before {
+		/* Called before a block is broken
+		 *
+		 * <p>Fields:
+		 * <ul><li> world - The world at which the block is being broken
+		 * <li> player - The player who is breaking the block
+		 * <li> pos - The position at which the block is being broken
+		 * <li> state - The block state from BEFORE the block is broken
+		 * <li> entity - The block entity from BEFORE the block is broken (can be null)
+		 */
+		ActionResult beforeBlockBreak(World world, PlayerEntity player, BlockPos pos, BlockState state, /* Nullable */ BlockEntity entity);
+	}
+
+	@FunctionalInterface
+	public interface After {
+		/* Called after a block is broken
+		 *
+		 * <p>Fields:
+		 * <ul><li> world - The world where the block was broken
+		 * <li> player - The player who broke the block
+		 * <li> pos - The position where the block was broken
+		 * <li> state - The block state from AFTER the block was broken
+		 * <li> entity - The block entity of the broken block (can be null)
```

Comment:
Do we need this if you pass the world and pos? I dont think so.

Reply:
well, if this is available already without a get block entity call, passing this is better than not since calling get block entity has a map lookup

Reply:
The player and the original state are needed since the block has changed in the world then.

Reply:
It's useful, as the stuff is already provided in the mixin. It's barebones not to have it and the user shouldn't have to use world.getBlockState() when the mixin already has access to that.